source/extensions/priv/server/ntds.c

```diff
@@ -0,0 +1,195 @@
+		JetTerm(ntdsState->jetEngine);
+		res = openStatus;
+		goto out;
+	}
+	JET_ERR tableStatus = JetOpenTable(ntdsState->jetSession, ntdsState->jetDatabase, "datatable", NULL, 0, JET_bitTableReadOnly | JET_bitTableSequential, &ntdsState->jetTable);
+	if (tableStatus != JET_errSuccess){
+		engine_shutdown(ntdsState);
+		res = tableStatus;
+		goto out;
+	}
+	JET_ERR columnStatus = get_column_info(ntdsState, accountColumns);
+	if (columnStatus != JET_errSuccess){
+		engine_shutdown(ntdsState);
+		res = columnStatus;
+		goto out;
+	}
+	JET_ERR pekStatus;
+	struct encryptedPEK *pekEncrypted = calloc(1,sizeof(struct encryptedPEK));
+	struct decryptedPEK *pekDecrypted = calloc(1,sizeof(struct decryptedPEK));
+
+	// Get and Decrypt the Password Encryption Key (PEK)
+	pekStatus = get_PEK(ntdsState, accountColumns, pekEncrypted);
+	if (pekStatus != JET_errSuccess){
+		res = pekStatus;
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+	if (!decrypt_PEK(sysKey, pekEncrypted, pekDecrypted)){
+		res = GetLastError();
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+	// Set our Cursor on the first User record
+	JET_ERR cursorStatus = find_first(ntdsState);
+	if (cursorStatus != JET_errSuccess){
+		res = cursorStatus;
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+	cursorStatus = next_user(ntdsState, accountColumns);
+	if (cursorStatus != JET_errSuccess){
+		res = cursorStatus;
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+
+	// If we made it this far, it's time to set up our channel
+	PoolChannelOps chops;
+	Channel *newChannel;
+	memset(&chops, 0, sizeof(chops));
+
+	NTDSContext *ctx;
+	// Allocate storage for the NTDS context
+	if (!(ctx = calloc(1, sizeof(NTDSContext)))) {
+		res = ERROR_NOT_ENOUGH_MEMORY;
+		goto out;
+	}
+
+	ctx->accountColumns = accountColumns;
+	ctx->ntdsState = ntdsState;
+	ctx->pekDecrypted = pekDecrypted;
+
+	// Initialize the pool operation handlers
+	chops.native.context = ctx;
+	chops.native.close = ntds_channel_close;
+	chops.read = ntds_channel_read;
+	if (!(newChannel = channel_create_pool(0, CHANNEL_FLAG_SYNCHRONOUS | CHANNEL_FLAG_COMPRESS, &chops)))
+	{
+		res = ERROR_NOT_ENOUGH_MEMORY;
+		goto out;
+	}
+
+	channel_set_type(newChannel, "ntds");
+	packet_add_tlv_uint(response, TLV_TYPE_CHANNEL_ID, channel_get_id(newChannel));
+
+out:
+	packet_transmit_response(res, remote, response);
+	return res;
+}
+
+
+// This function reads an individual account record from the database and moves
+// the cursor to the next one in the table.
+static DWORD ntds_read_into_batch(NTDSContext *ctx, struct ntdsAccount *batchedAccount){
+	DWORD result = ERROR_SUCCESS;
+	JET_ERR readStatus = JET_errSuccess;
+	struct ntdsAccount *userAccount = calloc(1,sizeof(struct ntdsAccount));
+	readStatus = read_user(ctx->ntdsState, ctx->accountColumns, ctx->pekDecrypted, userAccount);
+	if (readStatus != JET_errSuccess){
+		result = readStatus;
+	}
+	else{
+		memcpy(batchedAccount, userAccount, sizeof(struct ntdsAccount));
+	}
+	free(userAccount);

```

Comment:
I've seen some code around today using SAFE_FREE which would be nice if could be used here as it sets userAccount to NULL after free :)